### PR TITLE
Enable WebAssembly on Python 3

### PR DESCRIPTION
**Pull request by turbolent**
- On Python 3, convert Python `bytes` to JavaScript `ArrayBuffer`s 
- Add a `to_bytes` method to JavaScript objects, which converts `TypedArray`s to Python `bytes`
- Add a test for WebAssembly

### CODE REVIEW

**Comment by tbodt** [comment]
Neat! I'll take a look tomorrow.

File: tests/test_convert.py [hunk 1]
```diff
@@ -34,9 +34,6 @@ def test_convert_to_py(context):
                 " is called error."
     assert context.eval('"' + long_text + '"') == long_text
 
-    if sys.version_info.major >= 3:
-        assert context.eval('new Uint8Array([1, 2, 3])') == bytes([1, 2, 3])
-
```
Comment (tbodt) [inline tests/test_convert.py]: Why remove the test?

Comment (tbodt) [inline tests/test_convert.py]: Why remove the test?

File: v8py/jsobject.cpp [hunk 2]
```diff
@@ -13,6 +13,7 @@ PyTypeObject js_object_type = {
 };
 PyMethodDef js_object_methods[] = {
     {"__dir__", (PyCFunction) js_object_dir, METH_NOARGS, NULL},
+    {"to_bytes", (PyCFunction) js_object_to_bytes, METH_NOARGS, NULL},
```
Comment (tbodt) [inline v8py/jsobject.cpp]: This is problematic, it leaves no way to call a javascript method on the object called `to_bytes`.

Comment (tbodt) [inline v8py/jsobject.cpp]: This is problematic, it leaves no way to call a javascript method on the object called `to_bytes`.

File: v8py/convert.cpp [hunk 3]
```diff
@@ -141,27 +142,29 @@ Local<Value> js_from_py(PyObject *value, Local<Context> context) {
...
 #if PY_MAJOR_VERSION >= 3
+    if (PyBytes_Check(value)) {
+        Py_buffer view;
+        int error = PyObject_GetBuffer(value, &view, PyBUF_SIMPLE);
+        Local<ArrayBuffer> js_value = ArrayBuffer::New(isolate, view.len);
+        memcpy(js_value->GetContents().Data(), view.buf, view.len);
+        PyBuffer_Release(&view);
+        return hs.Escape(js_value);
+    }
+
```
Comment (tbodt) [inline v8py/convert.cpp]: What's the difference between this code and the code you deleted?

**Comment by tbodt** [comment]
Why not use bytearray instead of bytes? That works with both python 2 and python 3, and it's mutable so the data can actually be shared.

**Comment by turbolent** [comment]
Initially I just tried to pass `bytes` as such into V8 to instantiate a WebAssembly module (https://github.com/tbodt/v8py/pull/31/commits/ebd08f99a271e50414b23aeb91047b639e900860).
I also added the dual, to convert typed arrays back from V8 to Python `bytes` so a WebAssembly module's memory can be accessed from Python.

However, I then realized that the WebAssembly module's memory might want to be mutated, e.g. to pass data into the module, using [`TypedArray.set`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray/set). So instead, I removed the implicit conversion from V8 typed arrays to Python bytes in https://github.com/tbodt/v8py/pull/31/commits/de255482a7ef36cc6a765463cd4e585f645c35dd) and added the `to_bytes` method to still be able to do the conversion manually.

I guess to avoid the possible name-class for the `to_bytes` method, it could be exposed as a function from the v8py module (like `new`).

I agree that making `bytearray` work would be an even better solution, but I don't quite see how to implement that. Ideally, a slicing assignment to the bytearray in Python , e.g. `arr[1:3]`, would translate to a `TypedArray.set` call.

I looked at both the [C API for buffers](https://docs.python.org/3/c-api/buffer.html) and for [byte arrays](https://docs.python.org/dev/c-api/bytearray.html), but I'm not sure how too implement it.

I guess `__setitem__` could be implemented for `js_object`, and implemented as a `memcpy` to `ArrayBuffer.GetContents.Data`?

**Comment by tbodt** [comment]
The V8 API supports externalized array buffers, so no need to do all that. You can just create an array buffer backed by memory owned by a python bytearray. It should also be possible to create a python bytearray or memoryview or similar backed by an array buffer.

**Pull request closed without merging**